src/lib/comms/sol-bluetooth-impl-bluez.c

```diff
@@ -1581,6 +1640,613 @@ sol_bt_stop_scan(struct sol_bt_scan_pending *scan)
+{
+    struct context *ctx = userdata;
+
+    if (ctx->auth_conn)
+        sol_bt_conn_unref(ctx->auth_conn);
+
+    ctx->auth_conn = NULL;
+
+    ctx->agent_msg = sd_bus_message_unref(ctx->agent_msg);
+
+    return 0;
+}
+
+static const sd_bus_vtable agent_vtable[] = {
+    SD_BUS_VTABLE_START(0),
+    SD_BUS_METHOD("Release", NULL, NULL,
+        agent_release, SD_BUS_VTABLE_UNPRIVILEGED),
+    SD_BUS_METHOD("RequestPinCode", "o", NULL,
+        agent_request_pin_code, SD_BUS_VTABLE_UNPRIVILEGED),
+    SD_BUS_METHOD("DisplayPinCode", "os", NULL,
+        agent_display_pin_code, SD_BUS_VTABLE_UNPRIVILEGED),
+    SD_BUS_METHOD("RequestPasskey", "o", NULL,
+        agent_request_passkey, SD_BUS_VTABLE_UNPRIVILEGED),
+    SD_BUS_METHOD("DisplayPasskey", "ouq", NULL,
+        agent_display_passkey, SD_BUS_VTABLE_UNPRIVILEGED),
+    SD_BUS_METHOD("RequestConfirmation", "ou", NULL,
+        agent_request_confirmation, SD_BUS_VTABLE_UNPRIVILEGED),
+    SD_BUS_METHOD("RequestAuthorization", "o", NULL,
+        agent_request_authorization, SD_BUS_VTABLE_UNPRIVILEGED),
+    SD_BUS_METHOD("AuthorizeService", "os", NULL,
+        agent_authorize_service, SD_BUS_VTABLE_UNPRIVILEGED),
+    SD_BUS_METHOD("Cancel", NULL, NULL,
+        agent_cancel, SD_BUS_VTABLE_UNPRIVILEGED | SD_BUS_VTABLE_METHOD_NO_REPLY),
+    SD_BUS_VTABLE_END,
+};
+
+static const char *
+agent_to_capabities(const struct sol_bt_agent *agent)
+{
+    if (agent->passkey_entry && agent->passkey_display)
+        return "KeyboardDisplay";
+
+    if (agent->passkey_confirm && agent->passkey_display)
+        return "DisplayYesNo";
+
+    if (agent->passkey_entry)
+        return "KeyboardOnly";
+
+    if (agent->passkey_confirm)
+        return "DisplayOnly";
+
+    return "NoInputNoOutput";
+}
+
+static int
+register_agent_reply(sd_bus_message *reply, void *userdata, sd_bus_error *ret_error)
+{
+    struct context *ctx = userdata;
+    sd_bus *bus = sol_bus_client_get_bus(ctx->bluez);
+    const char *service = sol_bus_client_get_service(ctx->bluez);
+    const char *path = AGENT_PATH;
+    int r;
+
+    ctx->register_slot = sd_bus_slot_unref(ctx->register_slot);
+
+    if (sol_bus_log_callback(reply, userdata, ret_error))
+        return -EINVAL;
+
+    /* Don't really care if doesn't get to be the default agent */
+    r = sd_bus_call_method_async(bus, NULL, service, "/org/bluez",
+        "org.bluez.AgentManager1", "RequestDefaultAgent", sol_bus_log_callback, NULL,
+        "o", path);
+    SOL_INT_CHECK(r, < 0, r);
+
+    return 0;
+}
+
+SOL_API int
+sol_bt_register_agent(const struct sol_bt_agent *agent, void *data)
+{
+    struct context *ctx = &context;
+    sd_bus *bus = sol_bus_client_get_bus(ctx->bluez);
+    const char *service = sol_bus_client_get_service(ctx->bluez);
+    const char *path = AGENT_PATH;
+    const char *capabilities;
+    int r;
+
+    SOL_NULL_CHECK(agent, -EINVAL);
+
+    if (ctx->agent)
+        return -EALREADY;
```

Comment:
In `sol-ipm` we're using `-EEXIST` for this error, see https://github.com/solettaproject/soletta/blob/master/src/lib/io/sol-ipm.c#L254